### DISCUSSION TITLE
Event for "Menu Closed" ?

### DISCUSSION THREAD

**Discussion by httpete** [ariakit/ariakit#2423] [Questions] [answered]
Thank you for the continued progress of ariakit, we are using it heavily here and is huge piece of our stack.

I am implementing a "popup dialog" idea where I need to clear the values of an abandoned form. I see the new getState() stuff, but subscribing to "open" is leading to logic (like is this the first render, etc). 

Does ariakit have an event system I can connect to for this use? I really want to know when the menu is has just closed.

**Comment by diegohaz** [comment, marked as answer]
It's the same as before:

```jsx
useMenuStore({
  setOpen(open) {
    if (!open) {
      // on close...
    }
  }
});
```

More information: https://ariakit.org/guide/component-stores#state-setters

**Reply by httpete** [reply]
Thank you - To understand this though, you are passing a function to the setOpen property of the store, which is called internally? So I can think of the "setFoo" methods as event listeners?

**Reply by diegohaz** [reply]
I wouldn't call them event listeners because you can't register/unregister them as you would normally do with event listeners. But you can use them to call your event listeners.

You can think of them as state setters in a state object:

```js
const state = {
  open: false,
  setOpen(open) {
    this.open = open;
    // do something else
  },
};

state.setOpen(true);
```

Some examples:

- [Dialog with React Router](https://ariakit.org/examples/dialog-react-router)
- [Dialog with Next.js App Router](https://ariakit.org/examples/dialog-next-router)

**Reply by httpete** [reply]
Thank you.  It works perfectly. I use Zustand heavily and use their subscription / selector pattern. This is like you say more a setter.

**Reply by diegohaz** [reply]
Yeah! The store has some internal methods like `subscribe`, `sync`, and `syncBatch`, but they're not part of the public API right now.
